File: src/main/scala/com/vc/easy/L20.scala
package com.vc.easy

object L20 {
  def isValid(s: String): Boolean = {
    import scala.collection.mutable
    val st = new mutable.Stack[Char]()
    s.foreach({
      case '(' => st.push('(')
      case '{' => st.push('{')
      case '[' => st.push('[')
      case ')' => if(st.isEmpty || st.pop != '(') return false
      case '}' => if(st.isEmpty || st.pop != '{') return false
      case ']' => if(st.isEmpty || st.pop != '[') return false
    })
    st.isEmpty
  }
}

File: src/main/java/com/vc/easy/L20.java
package com.vc.easy;

import java.util.Stack;

class L20 {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();

        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) == ')') {
                if(st.isEmpty() || st.pop() != '(') return false;
            }
            else if(s.charAt(i) == ']') {
                if(st.isEmpty() || st.pop() != '[') return false;
            }
            else if(s.charAt(i) == '}') {
                if(st.isEmpty() || st.pop() != '{') return false;
            }
            else st.push(s.charAt(i));
        }
        return st.size() == 0;
    }
}
